### ISSUE TITLE
Unified_inventory mod crashes when clicking on the "building materials" icon

### ISSUE THREAD

**Issue by dennisjenkins75** [minetest-mods/unified_inventory#245]
This is similar to issue #243, but occurs when clicking on the "Building Materials" icon.  (Clicking on the iron ore works fine).

Observed on server "EdenLost", running minetest-5.8.0.  This is a public server, so bisecting the list of installed mods would require significant effort (taking it offline, making a backup, disabling logins, experimenting, possibly restoring a backup, etc...)  I am hoping that between the context in issue #243 and the stack trace below a simple fix (check for nil before using it as a table index) would be sufficient.

```
1hit@EdgyNet:~/worlds/world/worldmods/unified_inventory$ git rev-parse HEAD
921a6d76ee16f3633cf6122800da37776557b20a

1hit@EdgyNet:~$ ./bin/crash-report.py < ./logs/2024-02-24.txt 
ServerError: AsyncErr: Lua: Runtime error from mod 'unified_inventory' in callback on_playerReceiveFields(): ...it/worlds/world/worldmods/unified_inventory/internal.lua:387: attempt to index local 'def' (a nil value)
stack traceback:
        ...it/worlds/world/worldmods/unified_inventory/internal.lua:387: in function 'ffilter'
        ...it/worlds/world/worldmods/unified_inventory/internal.lua:417: in function 'apply_filter'
        ...t/worlds/world/worldmods/unified_inventory/callbacks.lua:114: in function <...t/worlds/world/worldmods/unified_inventory/callbacks.lua:91>
        .../opt/minetest/share/minetest/builtin/common/register.lua:26: in function <.../opt/minetest/share/minetest/builtin/common/register.lua:12>
```

I modified the code around the crash site to log the `name` that fails to be found in `registered_items` and also return.  The crash is now gone, and I have some data:

`internal.lua:384`:
```
                ffilter = function(name)
                        local def = registered_items[name]
  if not def then
    minetest.log("error", "[unified_inventory] Failed to resolve registered_item " .. name)
    return false
  end
```

Logfile:
```
2024-02-24 19:26:15: ERROR[Server]: [unified_inventory] Failed to resolve registered_item stairs:stair_inner_snowblock
2024-02-24 19:26:15: ERROR[Server]: [unified_inventory] Failed to resolve registered_item stairs:stair_outer_snowblock
```

**Comment by SmallJoker** [comment]
This should fix it for once and all: (feedback is welcome)
```diff
diff --git a/api.lua b/api.lua
index f2f40c2..d748167 100644
--- a/api.lua
+++ b/api.lua
@@ -184,6 +184,18 @@ minetest.after(0.01, function()
                ui.crafts_for.recipe[outputitemname] = new_recipe_list
        end
 
+       -- Remove unknown items from all categories
+       for cat_name, cat_def in pairs(ui.registered_category_items) do
+               for itemname, exists in pairs(cat_def) do
+                       if exists and not minetest.registered_items[itemname] then
+                               minetest.log("warning", "[unified_inventory] Removed item '"
+                                       .. itemname .. "' from category '" .. cat_name
+                                       .. "'. Reason: item not registered")
+                               cat_def[itemname] = nil
+                       end
+               end
+       end
+
        for _, callback in ipairs(ui.initialized_callbacks) do
                callback()
        end
diff --git a/internal.lua b/internal.lua
index daa3ba1..3caa43a 100644
--- a/internal.lua
+++ b/internal.lua
@@ -368,6 +368,10 @@ function ui.apply_filter(player, filter, search_dir)
                local groups = lfilter:sub(7):split(",")
                ffilter = function(name)
                        local def = registered_items[name]
+                       if not def then
+                               return false
+                       end
+
                        for _, group in ipairs(groups) do
                                if not def.groups[group]
                                or def.groups[group] <= 0 then
@@ -383,6 +387,10 @@ function ui.apply_filter(player, filter, search_dir)
 
                ffilter = function(name)
                        local def = registered_items[name]
+                       if not def then
+                               return false
+                       end
+
                        local lname = string.lower(name)
                        local ldesc = string.lower(def.description)
                        local llocaldesc = minetest.get_translated_string
```

**Comment by Andarius68** [comment]
Yep, SmallJoker's fix worked for me. Thank you for that! :)

**Comment by SmallJoker** [comment]
Thank you. Fixed in b5de18b
